diff --git a/gradle/libs.versions.toml b/gradle/libs.versions.toml
index c765756e..51017d64 100644
--- a/gradle/libs.versions.toml
+++ b/gradle/libs.versions.toml
@@ -58,4 +58,5 @@ robolectric = { group = "org.robolectric", name = "robolectric", version.ref = "
 
 [plugins]
 android-application = { id = "com.android.application", version.ref = "android-gradle-plugin" }
+android-library = { id = "com.android.library", version.ref = "android-gradle-plugin" }
 kotlin-android = { id = "org.jetbrains.kotlin.android", version.ref = "kotlin" }
\ No newline at end of file
diff --git a/main/build.gradle.kts b/main/build.gradle.kts
index 0cb27409..1f3bef00 100644
--- a/main/build.gradle.kts
+++ b/main/build.gradle.kts
@@ -8,10 +8,11 @@ import org.gradle.api.tasks.TaskProvider
  */
 
 plugins {
-    alias(libs.plugins.android.application)
-    id("checkstyle")
+    alias(libs.plugins.android.library)
 }
 
+val isEmbeddedLibrary = project.name == "openvpn"
+
 
 fun obtainTestBuildType(): String {
     var result = "debug";
@@ -36,10 +37,6 @@ android {
 
     defaultConfig {
         minSdk = 23
-        targetSdk = 37
-        //targetSdkPreview = "UpsideDownCake"
-        versionCode = 219
-        versionName = "0.7.64"
         externalNativeBuild {
             cmake {
                 //arguments+= "-DCMAKE_VERBOSE_MAKEFILE=1"
@@ -130,19 +127,20 @@ android {
 
         create("ovpn2") {
             dimension = "ovpnimpl"
-            versionNameSuffix = "-o2"
             buildConfigField("boolean", "openvpn3", "false")
         }
     }
 
     buildTypes {
         getByName("release") {
-            if (project.hasProperty("icsopenvpnDebugSign")) {
-                logger.warn("property icsopenvpnDebugSign set, using debug signing for release")
-                signingConfig = android.signingConfigs.getByName("debug")
-            } else {
-                productFlavors["ovpn23"].signingConfig = signingConfigs.getByName("release")
-                productFlavors["ovpn2"].signingConfig = signingConfigs.getByName("releaseOvpn2")
+            if (!isEmbeddedLibrary) {
+                if (project.hasProperty("icsopenvpnDebugSign")) {
+                    logger.warn("property icsopenvpnDebugSign set, using debug signing for release")
+                    signingConfig = android.signingConfigs.getByName("debug")
+                } else {
+                    productFlavors["ovpn23"].signingConfig = signingConfigs.getByName("release")
+                    productFlavors["ovpn2"].signingConfig = signingConfigs.getByName("releaseOvpn2")
+                }
             }
         }
     }
@@ -154,12 +152,14 @@ android {
         sourceCompatibility = JavaVersion.VERSION_17
     }
 
-    splits {
-        abi {
-            isEnable = true
-            reset()
-            include("x86", "x86_64", "armeabi-v7a", "arm64-v8a")
-            isUniversalApk = true
+    if (!isEmbeddedLibrary) {
+        splits {
+            abi {
+                isEnable = true
+                reset()
+                include("x86", "x86_64", "armeabi-v7a", "arm64-v8a")
+                isUniversalApk = true
+            }
         }
     }
 
@@ -169,34 +169,6 @@ android {
         }
     }
 
-    bundle {
-        codeTransparency {
-            signing {
-                val keystoreTPFile: String? by project
-                storeFile = keystoreTPFile?.let { file(it) }
-                val keystoreTPPassword: String? by project
-                storePassword = keystoreTPPassword
-                val keystoreTPAliasPassword: String? by project
-                keyPassword = keystoreTPAliasPassword
-                val keystoreTPAlias: String? by project
-                keyAlias = keystoreTPAlias
-
-                if (keystoreTPFile?.isEmpty()
-                        ?: true
-                ) println("keystoreTPFile not set, disabling transparency signing")
-                if (keystoreTPPassword?.isEmpty()
-                        ?: true
-                ) println("keystoreTPPassword not set, disabling transparency signing")
-                if (keystoreTPAliasPassword?.isEmpty()
-                        ?: true
-                ) println("keystoreTPAliasPassword not set, disabling transparency signing")
-                if (keystoreTPAlias?.isEmpty()
-                        ?: true
-                ) println("keyAlias not set, disabling transparency signing")
-
-            }
-        }
-    }
 }
 
 var swigcmd = "swig"
